### PR TITLE
`<algorithm>`: Implement batched random integer generation for `shuffle()`

### LINKED ISSUE #5736: `<algorithm>`: Investigate further optimizations for `shuffle()` and `sample()`
After seeing #5735, @lemire pointed me to [Batched Ranged Random Integer Generation](https://arxiv.org/pdf/2408.06213) published in Aug 2024 by Nevin Brackett-Rozinsky and himself, which sounds very promising for the STL:

> Pseudorandom values are often generated as 64-bit binary words. These random words need to be converted into ranged values without statistical bias. We present an efficient algorithm to generate multiple independent uniformly-random bounded integers from a single uniformly-random binary word, without any bias. In the common case, our method uses one multiplication and no division operations per value produced. In practice, our algorithm can more than double the speed of unbiased random shuffling for small to moderately large arrays.

He expects that it would probably take just a few lines of code.

### PR DESCRIPTION
**Pull request by FranciscoThiesen**
## Summary
- Implement batched random number generation for shuffle based on Brackett-Rozinsky & Lemire's paper
- For URNGs with full 64-bit range (like mt19937_64), generate 2 bounded random integers from a single 64-bit random value
- Uses Lemire's nearly-divisionless algorithm for rejection sampling

## Implementation Details
- Added _Batched_rng_from_urng struct that extracts batches of bounded random integers
- Added _Random_shuffle_batched function implementing Fisher-Yates with batched RNG
- Modified shuffle() and ranges::shuffle to use batched version for eligible URNGs
- Compile-time dispatch based on URNG characteristics (full 64-bit range)

## Test Plan
- [x] Existing shuffle tests pass
- [x] Added permutation correctness tests for both 64-bit (batched) and 32-bit (non-batched) paths
- [x] Added edge case tests (empty, single element, 2, 3, 4, and 10000 elements)

Resolves #5736

### CODE REVIEW

File: stl/inc/algorithm [hunk 1]
```diff
@@ -6080,6 +6096,92 @@ namespace ranges {
...
+
+template <class _Diff, class _Urng>
+struct _Batched_rng_from_urng {
+    // Batched random generation is only beneficial for 64-bit RNGs with full range.
+    // It requires the RNG to produce values in [0, 2^64 - 1].
+    using _Urng_result = _Invoke_result_t<_Urng&>;
+
+    static constexpr bool _Has_full_64bit_range = sizeof(_Urng_result) >= sizeof(uint64_t)
+                                               && is_unsigned_v<_Urng_result> && (_Urng::min) () == 0
+                                               && (_Urng::max) () == (numeric_limits<uint64_t>::max) ();
```
Comment (frederick-vs-ja) [inline stl/inc/algorithm]: Note that `is_unsigned_v<T>` can only be `true` for integer types. I think on MSVC ABI only `unsigned long long` (and its cv-qualified versions) satisfies such conditions.

Comment (frederick-vs-ja) [reply stl/inc/algorithm]: Oops, I was wrong on some conditions. Perhaps we should move this out of the class body to reduce instantiation.

Perhaps it's better to introduce
```C++
template <class _Urng>
constexpr bool _Urng_has_full_64bit_range = is_same_v<_Invoke_result_t<_Urng>, uint64_t> && (_Urng::min) () == 0
                                         && (_Urng::max) () == _STD _Max_limit<uint64_t>();
```

**Comment by FranciscoThiesen** [comment]
> @FranciscoThiesen please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@microsoft-github-policy-service agree [company="Microsoft"]

**Comment by FranciscoThiesen** [comment]
@microsoft-github-policy-service agree company="Microsoft"

File: stl/inc/algorithm (hunk 1 above)
Comment (FranciscoThiesen) [reply stl/inc/algorithm]: Done.

File: stl/inc/algorithm [hunk 2]
```diff
@@ -6080,6 +6096,91 @@ namespace ranges {
...
+        }
+
+        return static_cast<_Diff>(_Product._Word[1]);
+    }
+
+    // Generate two bounded random values from a single 64-bit random word.
+    // The bounds are (n+1) and n for Fisher-Yates shuffle positions _Target_index and _Target_index-1.
+    void _Batch_2(_Diff* _Results, _Diff _Bound1, _Diff _Bound2) {
+        const uint64_t _B1            = static_cast<uint64_t>(_Bound1);
+        const uint64_t _B2            = static_cast<uint64_t>(_Bound2);
```
Comment (StephanTLavavej) [inline stl/inc/algorithm]: Naming nitpick: We conventionally avoid macro-like identifiers such as underscore-capital (`_N`) and underscore-capital-number (`_B1`) because they contain no lowercase letters.

File: stl/inc/algorithm [hunk 3]
```diff
@@ -6439,20 +6620,37 @@ namespace ranges {
         _STATIC_CALL_OPERATOR _It operator()(_It _First, _Se _Last, _Urng&& _Func) _CONST_CALL_OPERATOR {
             _STD _Adl_verify_range(_First, _Last);
 
-            _Rng_from_urng_v2<iter_difference_t<_It>, remove_reference_t<_Urng>> _RngFunc(_Func);
-            auto _UResult = _Shuffle_unchecked(
-                _RANGES _Unwrap_iter<_Se>(_STD move(_First)), _RANGES _Unwrap_sent<_It>(_STD move(_Last)), _RngFunc);
+            using _Urng0 = remove_reference_t<_Urng>;
+            using _Diff  = iter_difference_t<_It>;
```
Comment (StephanTLavavej) [inline stl/inc/algorithm]: `_Diff` is only needed in the `else` branch. Occurs below.

File: stl/inc/algorithm [hunk 4]
```diff
@@ -6423,11 +6524,91 @@ void _Random_shuffle1(_RanIt _First, _RanIt _Last, _RngFn& _RngFunc) {
     }
 }
 
+// Batched shuffle implementation for 64-bit URNGs with full range.
+// Uses batched random generation to reduce RNG calls.
+template <class _RanIt, class _Urng>
+void _Random_shuffle_batched(_RanIt _First, _RanIt _Last, _Urng& _Func) {
+    // shuffle [_First, _Last) using batched random generation
+    _STD _Adl_verify_range(_First, _Last);
+    auto _UFirst      = _STD _Get_unwrapped(_First);
```
Comment (StephanTLavavej) [inline stl/inc/algorithm]: `_UFirst` could be const.

File: stl/inc/algorithm [hunk 5]
```diff
@@ -6080,6 +6096,91 @@ namespace ranges {
...
+    static constexpr uint64_t _Bound_for_batch_2 = 4294967296; // 2^32, for batch of 2
+
+    _Urng& _Ref;
+
+    explicit _Batched_rng_from_urng(_Urng& _Func) noexcept : _Ref(_Func) {}
+
+    // Generate a single bounded random value in [0, _Bound) using Lemire's method.
+    _NODISCARD _Diff _Single_bounded(_Diff _Bound) {
+        _Unsigned128 _Product{
+            _Base128::_UMul128(static_cast<uint64_t>(_Ref()), static_cast<uint64_t>(_Bound), _Product._Word[1])};
```
Comment (StephanTLavavej) [inline stl/inc/algorithm]: `_Batched_rng_from_urng` requires `_Urng_has_full_64bit_range`, so we're guaranteed that `is_same_v<_Invoke_result_t<_Urng&>, uint64_t>` and thus `static_cast<uint64_t>(_Ref())` is unnecessary.

We should add internal static asserts and then remove the casts.

File: stl/inc/algorithm [hunk 6]
```diff
@@ -6080,6 +6096,91 @@ namespace ranges {
...
+    static constexpr uint64_t _Bound_for_batch_4 = 302; // (302)^4 = 8,319,430,096 < 2^64
+    static constexpr uint64_t _Bound_for_batch_3 = 2642; // (2642)^3 = 18,454,249,288 < 2^64
+    static constexpr uint64_t _Bound_for_batch_2 = 4294967296; // 2^32, for batch of 2
+
+    _Urng& _Ref;
+
+    explicit _Batched_rng_from_urng(_Urng& _Func) noexcept : _Ref(_Func) {}
+
+    // Generate a single bounded random value in [0, _Bound) using Lemire's method.
+    _NODISCARD _Diff _Single_bounded(_Diff _Bound) {
```
Comment (StephanTLavavej) [inline stl/inc/algorithm]: We could add const to parameters. Occurs below.

File: stl/inc/algorithm [hunk 7]
```diff
@@ -6080,6 +6096,91 @@ namespace ranges {
...
+                _Leftover = _Product._Word[0];
+            }
+        }
+
+        return static_cast<_Diff>(_Product._Word[1]);
+    }
+
+    // Generate two bounded random values from a single 64-bit random word.
+    // The bounds are (n+1) and n for Fisher-Yates shuffle positions _Target_index and _Target_index-1.
+    void _Batch_2(_Diff* _Results, _Diff _Bound1, _Diff _Bound2) {
```
Comment (StephanTLavavej) [inline stl/inc/algorithm]: `_Batch_2()` is always called with `_Diff _Offsets[2];`. It would be clearer and safer to take a reference-to-array.

File: stl/inc/algorithm [hunk 8]
```diff
@@ -6080,6 +6096,91 @@ namespace ranges {
...
+
+    explicit _Batched_rng_from_urng(_Urng& _Func) noexcept : _Ref(_Func) {}
+
+    // Generate a single bounded random value in [0, _Bound) using Lemire's method.
+    _NODISCARD _Diff _Single_bounded(_Diff _Bound) {
+        _Unsigned128 _Product{
+            _Base128::_UMul128(static_cast<uint64_t>(_Ref()), static_cast<uint64_t>(_Bound), _Product._Word[1])};
+        auto _Leftover = _Product._Word[0];
+
+        if (_Leftover < static_cast<uint64_t>(_Bound)) {
```
Comment (StephanTLavavej) [inline stl/inc/algorithm]: `static_cast<uint64_t>(_Bound)` is verbosely repeated 5 times and would benefit from being extracted.

File: stl/inc/algorithm [hunk 9]
```diff
@@ -6080,6 +6096,91 @@ namespace ranges {
...
+template <class _Diff, class _Urng>
+struct _Batched_rng_from_urng {
+
+    // Threshold bounds for batch sizes based on array size.
+    // These are derived from the paper to minimize expected cost per random value.
+    // Batch size k requires product of k consecutive bounds <= 2^64.
+    static constexpr uint64_t _Bound_for_batch_6 = 21; // (21)^6 = 85,766,121 < 2^64 / (very conservative)
+    static constexpr uint64_t _Bound_for_batch_5 = 73; // (73)^5 = 2,073,071,593 < 2^64
+    static constexpr uint64_t _Bound_for_batch_4 = 302; // (302)^4 = 8,319,430,096 < 2^64
+    static constexpr uint64_t _Bound_for_batch_3 = 2642; // (2642)^3 = 18,454,249,288 < 2^64
```
Comment (StephanTLavavej) [inline stl/inc/algorithm]: These bounds are all unused. Two of the comments are mathematically incorrect:

```
C:\Temp>py
Python 3.14.3 (tags/v3.14.3:323c59a, Feb  3 2026, 16:04:56) [MSC v.1944 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> 302**4
8318169616
>>> 2642**3
18441593288
```

File: tests/std/tests/P0896R4_ranges_alg_shuffle/test.cpp [hunk 10]
```diff
@@ -72,8 +75,125 @@ void test_urbg() { // COMPILE-ONLY
...
 }
 
+// Test that shuffle produces a valid permutation for various sizes.
+// This exercises both the batched path (for 64-bit RNGs) and the fallback path.
+void test_shuffle_permutation() {
+    // Test with 64-bit generator (batched random path)
+    {
+        vector<int> v(100);
+        iota(v.begin(), v.end(), 0);
+        vector<int> original = v;
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_alg_shuffle/test.cpp]: We could extract this and add const to the `original` vectors.

File: tests/std/tests/P0896R4_ranges_alg_shuffle/test.cpp [hunk 11]
```diff
@@ -72,8 +75,125 @@ void test_urbg() { // COMPILE-ONLY
...
+
+        shuffle(v.begin(), v.end(), gen64);
+
+        // Verify it's still a permutation
+        vector<int> sorted_v = v;
+        sort(sorted_v.begin(), sorted_v.end());
+        assert(sorted_v == original);
+    }
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_alg_shuffle/test.cpp]: We recreate `v` every time, so there's no need to copy it before sorting, as `test_shuffle_edge_cases()` already demonstrates.

File: tests/std/tests/P0896R4_ranges_alg_shuffle/test.cpp [hunk 12]
```diff
@@ -72,8 +75,125 @@ void test_urbg() { // COMPILE-ONLY
...
+        vector<int> v        = {1, 2, 3, 4};
+        vector<int> original = v;
+        shuffle(v.begin(), v.end(), gen64);
+        sort(v.begin(), v.end());
+        assert(v == original);
+    }
+
+    // Large array to ensure batching is effective
+    {
+        vector<int> v(10000);
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_alg_shuffle/test.cpp]: These are highly systematic and could be centralized. Additionally, I would feel better with a large but odd number.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo. Please notify me if any further changes are pushed, otherwise no action is required.

**Comment by StephanTLavavej** [comment]
:rocket: :game_die: :heart_eyes_cat:
===

**Comment by lemire** [comment]
@StephanTLavavej My employer had be take a training course where we were told that emojis are unprofessional and should be avoided.

🔥 🚀  Go!
